src/Storages/RabbitMQ/StorageRabbitMQ.cpp

```diff
@@ -0,0 +1,551 @@
+#include <Storages/RabbitMQ/RabbitMQHandler.h>
+#include <Storages/StorageFactory.h>
+#include <Storages/StorageMaterializedView.h>
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/trim.hpp>
+#include <Common/Exception.h>
+#include <Common/Macros.h>
+#include <Common/config_version.h>
+#include <Common/setThreadName.h>
+#include <Common/typeid_cast.h>
+#include <common/logger_useful.h>
+#include <Common/quoteString.h>
+#include <Common/parseAddress.h>
+#include <Processors/Sources/SourceFromInputStream.h>
+#include <amqpcpp.h>
+
+namespace DB
+{
+
+enum
+{
+    Connection_setup_sleep = 200,
+    Connection_setup_retries_max = 1000
+};
+
+static const auto RESCHEDULE_MS = 500;
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int BAD_ARGUMENTS;
+}
+
+
+StorageRabbitMQ::StorageRabbitMQ(
+        const StorageID & table_id_,
+        Context & context_,
+        const ColumnsDescription & columns_,
+        const String & host_port_,
+        const Names & routing_keys_,
+        const String & exchange_name_,
+        const String & format_name_,
+        char row_delimiter_,
+        const String & exchange_type_,
+        size_t num_consumers_,
+        size_t num_queues_,
+        const bool use_transactional_channel_)
+        : IStorage(table_id_)
+        , global_context(context_.getGlobalContext())
+        , rabbitmq_context(Context(global_context))
+        , routing_keys(global_context.getMacros()->expand(routing_keys_))
+        , exchange_name(exchange_name_)
+        , format_name(global_context.getMacros()->expand(format_name_))
+        , row_delimiter(row_delimiter_)
+        , num_consumers(num_consumers_)
+        , num_queues(num_queues_)
+        , exchange_type(exchange_type_)
+        , use_transactional_channel(use_transactional_channel_)
+        , log(&Poco::Logger::get("StorageRabbitMQ (" + table_id_.table_name + ")"))
+        , semaphore(0, num_consumers_)
+        , login_password(std::make_pair(
+                    rabbitmq_context.getConfigRef().getString("rabbitmq_username", "root"),
+                    rabbitmq_context.getConfigRef().getString("rabbitmq_password", "clickhouse")))
+        , parsed_address(parseAddress(global_context.getMacros()->expand(host_port_), 5672))
+{
+    loop = std::make_unique<uv_loop_t>();
+    uv_loop_init(loop.get());
+
+    event_handler = std::make_unique<RabbitMQHandler>(loop.get(), log);
+    connection = std::make_unique<AMQP::TcpConnection>(event_handler.get(), AMQP::Address(parsed_address.first, parsed_address.second, AMQP::Login(login_password.first, login_password.second), "/"));
+
+    size_t cnt_retries = 0;
+    while (!connection->ready() && ++cnt_retries != Connection_setup_retries_max)
+    {
+        uv_run(loop.get(), UV_RUN_NOWAIT);
+        std::this_thread::sleep_for(std::chrono::milliseconds(Connection_setup_sleep));
```

Comment:
Race condition in AMQCPP triggered here: (write)[https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/master/src/linux_tcp/tcpresolver.h#L100] , (read)[https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/master/src/linux_tcp/tcpresolver.h#L200]. Strange async code  https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/master/src/linux_tcp/tcpresolver.h#L166.